lib/core/test/data/Cardano/Wallet/Api/ApiEpochNumber.json

```diff
@@ -1,15 +0,0 @@
-{
-    "seed": 3314543689123684722,
-    "samples": [
-        "1617893896",
-        "258954262",
-        "latest",
-        "latest",
-        "1310519342",
-        "latest",
-        "42220161",
-        "latest",
-        "1542634324",
```

Comment:
This golden should be *changed* and not *deleted*?

Reply:
I deleted it because I moved validation downstream, as described in the above comment

Reply:
> I moved validation downstream

As I see it, the JSON does the same validation/parsing, but the actual failure (`throwE`/`fail`) is delayed to later.

The `jsonRoundtripAndGolden` ensure the custom json instances… roundtrips… and doesn't change drastically. 

I do not understand why you removed it 🤔

Reply:
I wanted to capture it on Server level, use in `ErrGetNetworkParametersInvalidValue` and emit err400. Otherwise it would return earlier - just when json validation and could not (easily - maybe not true) embrace it and treat equally as `ErrGetNetworkParametersNotSuchEpochNumber`

Reply:
So I extended : 
```
data ApiEpochNumber =
    ApiEpochNumberLatest | ApiEpochNumber EpochNo | ApiEpochNumberInvalid Text
```
and funneled text that do not satisfy "latest" or is not convertible to `Word31` to `ApiEpochNumberInvalid`. And if so I can on `Server` level handle errors, control error code, and err msg (this of course can be done also as it was done in the previous PR in instances )

Reply:
Yes, but

1) `ApiEpochNumberInvalid Text` seems strange to me. I guessed the problem it is solving is not new to our wallet, but it looked like a new solution.
2) I don't see how the existence of the roundtrip is related to this. I.e. why is the move of error control an argument for removing the roundtrip? It doesn't hurt to just keep it?

I looked in to this a bit.

Previously you used `fail` in `FromText ApiEpochNumber`. I noticed the endpoint would then `err500 "Something bad happened"` if provided an invalid parameter.

I suspect this is due to `fail` throwing an exception:
```Haskell
(fail "no") :: (Either String Int)
*** Exception: no
```

I noticed that existing endpoints ending with path parameters would not be _not found_ when the path parameter would be incorrect:

e.g. 

```bash
http delete http://localhost:8090/v2/wallets/1efcf74dfc539599f6fa3343859c172a325abbc7/transactions/1423856bc91c49e928f6f30f4e8d665d53eb4ab6028bd0ac971809d514c92db
HTTP/1.1 404 Not Found
Content-Type: application/json;charset=utf-8
Date: Tue, 28 Jan 2020 10:33:12 GMT
Server: Warp/3.2.27
Transfer-Encoding: chunked

{
    "code": "not_found",
    "message": "I couldn't find the requested endpoint. If the endpoint contains path parameters, please ensure they are well-formed, otherwise I won't be able to route them correctly."
}
```

Though doing so here would be against the U/S as you said.

However, I tried replacing your `ApiEpochNumberInvalid` with `Left $ TextDecodingError`, and somehow it actually _does_ return a `err400`:

```bash
http http://localhost:8090/v2/network/parameters/late
HTTP/1.1 400 Bad Request
Content-Type: application/json;charset=utf-8
Date: Tue, 28 Jan 2020 10:42:58 GMT
Server: Warp/3.2.27
Transfer-Encoding: chunked

{
    "code": "bad_request",
    "message": ""
}
```

I am not sure why, or from where the `bad_request` or `message` comes from.

🤷‍♂

Reply:
So seems a bit odd to me, but user-facing, it is a really nice experience the way you've done it.

Still, I'm wondering about this:

> I don't see how the existence of the roundtrip is related to this. I.e. why is the move of error control an argument for removing the roundtrip? It doesn't hurt to just keep it?

Reply:
no, we can resurrect it if you want. I thought, that when we move validation downstream, away from json decoder, then it is not needed

Reply:
BUT, as you can see we do not need ToJSON and FromJSON instances of ApiEpochNumber anymore (when we went downstream), so why to add them? Just for testing - I thought that maybe it is better not to bloat code base ;-)

Reply:
I'd vote for keeping the tests for:
 - making sure the system still passes those roundtrips after your changes.
 - as an additional guard for any future changes in the area.

(Unless there are now similar tests elsewhere?)

Reply:
added textual roundtrip https://github.com/input-output-hk/cardano-wallet/pull/1306/commits/8d7e850f3d41a7ac784430fed4b267bc703af31f